File: online/LeetCodeCN/1~100/70.cs
// 爬楼梯 https://leetcode-cn.com/problems/climbing-stairs/
// DP

public class Solution {
    private Dictionary<int, int> dp = new Dictionary<int, int>();

    public Solution() {
        dp.Add(1, 1);
        dp.Add(2, 2);
    }

    public int ClimbStairs(int n) {
        int result = 0;
        if (!dp.TryGetValue(n, out result)) {
            result = ClimbStairs(n-1) + ClimbStairs(n-2);
            dp.Add(n, result);
        }
        return result;
    }
}

File: online/LeetCodeCN/1~100/70.cpp
// 爬楼梯 https://leetcode-cn.com/problems/climbing-stairs/
// DP

class Solution {
    unordered_map<int, int> dp;

public:
    Solution() {
        dp[1] = 1;
        dp[2] = 2;
    }

    int climbStairs(int n) {
        return dp[n] ? dp[n] : (dp[n] = climbStairs(n-1)+ climbStairs(n-2));
    }
};
